src/dmd/constfold.d

```diff
@@ -1043,7 +1043,7 @@ UnionExp Cast(const ref Loc loc, Type type, Type to, Expression e1)
     Type typeb = type.toBasetype();
     //printf("Cast(type = %s, to = %s, e1 = %s)\n", type.toChars(), to.toChars(), e1.toChars());
     //printf("\te1.type = %s\n", e1.type.toChars());
-    if (e1.type.equals(type) && type.equals(to))
+    if ((e1.type.toBasetype.equals(type) || e1.type.equals(type)) && type.equals(to))
     {
         emplaceExp!(UnionExp)(&ue, e1);
```

Comment:
This will create a wrongly typed expression when `e1.type.toBasetype() != e1.type` and `e1.type.toBasetype.equals(type)` is true.

Reply:
Can you please provide an example where this is the case? The scenario you pointed out is also the scenario from the issue, where we have an enum E : char[4] and it won't cast to a char[4], although it's already a char[4] :). In that case e1.type is E, but I still wanted to enter in the first if block, because E and char[4] are the same thing, and that's why I added the condition (e1.type = E, e1.type.toBasetype = char[4]) and hence the expression will be casted. In most of the cases, the condition won't change anything.
Thank you.

Reply:
Don't have a specific code at hand now, will post an example if I found one.

The problem with the current state is that it breaks the contract of this method:
> type: type to paint the result

Should be resolvable by simply painting the emplaced expression in the aforementioned case.